### PR TITLE
Rather than calculate stats on demand when stale, backfill the cache

### PR DESCRIPTION
**Pull request by russelldb**
Always server the stats that are in the cache, no matter how old they are.
Add a timestamp to the stats so consumers know how stale they are.
Fill the cache continuously in the background.

See https://github.com/basho/riak_kv/issues/505 and https://github.com/basho/riak_kv/issues/506 for graphs, bench configs etc.

### CODE REVIEW

File: src/riak_core_stat_cache.erl [hunk 1]
```diff
@@ -18,14 +18,20 @@
 %%
 %% -------------------------------------------------------------------
 
+%% @ doc fetches stats for registered modules and stores them
```
Comment (rzezeski) [inline src/riak_core_stat_cache.erl]: I think this needs to be `@doc` ;)

File: src/riak_core_stat_cache.erl [hunk 2]
```diff
@@ -87,29 +101,42 @@ handle_call({get_stats, App}, From, State0=#state{apps=Apps, active=Active0, tab
...
-                            {reply, {ok, Stats, TS}, State0}
+                            FreshnessStat = make_freshness_stat(App, TS),
+                            {reply, {ok, [FreshnessStat | Stats], TS}, State0}
                     end
             end,
     Reply;
+handle_call({clear, App}, _From, State=#state{apps=Apps, tab=Tab}) ->
+    Reply = case registered(App, Apps) of
+                false ->
+                    {reply, ?ENOTREG(App), State};
```
Comment (rzezeski) [inline src/riak_core_stat_cache.erl]: I could go either way on this but perhaps just return `ok` in this case too?  The benefit I see is less chance of a `badmatch` error.  If an app doesn't exist then it's already cleared, right?

File: src/riak_core_stat_cache.erl [hunk 3]
```diff
@@ -136,10 +167,19 @@ code_change(_OldVsn, State, _Extra) ->
...
+    {make_freshness_stat_name(App), TS}.
+
+make_freshness_stat_name(App) ->
+    list_to_atom(atom_to_list(App) ++ "_stat_ts").
+
+register_mod(App, Mod, Fun, Args, RefreshRateMillis, Apps) ->
     folsom_metrics:new_histogram({?MODULE, Mod}),
     folsom_metrics:new_meter({?MODULE, App}),
-    orddict:store(App, {Mod, Fun, Args, TTL}, Apps).
+    orddict:store(App, {Mod, Fun, Args, RefreshRateMillis}, Apps).
```
Comment (rzezeski) [inline src/riak_core_stat_cache.erl]: I doubt it matters at all in terms of performance but it's interesting that there are 3 different data structures representing the Mod/Fun/Arg/RefreshRate data.
1. Here you store it as `{M, F, A, RR}`
2. Each is passed to `register_mod` as its own argument
3. The register `handle_call` has `{M, F, A}` as and `RR`

Maybe just pass around as `{{M, F, A}, RR}` everywhere to be consistent?  I dunno, I'm just nitpicking and at end of day shouldn't really matter.

File: src/riak_core_stat_cache.erl [hunk 4]
```diff
@@ -87,29 +101,42 @@ handle_call({get_stats, App}, From, State0=#state{apps=Apps, active=Active0, tab
     Reply = case registered(App, Apps) of
                 false ->
                     {reply, ?ENOTREG(App), State0};
-                {true, {M, F, A, TTL}} ->
-                    case cache_get(App, Tab, TTL) of
-                        No when No == miss; No == stale ->
+                {true, {M, F, A, _RefreshRateMillis}} ->
+                    case cache_get(App, Tab) of
+                        No when No == miss ->
```
Comment (rzezeski) [inline src/riak_core_stat_cache.erl]: Why not just `miss ->`?

Comment (russelldb) [inline src/riak_core_stat_cache.erl]: Sorry, used to be `miss` or `stale`, quick fixed it. Will change.

File: src/riak_core_stat_cache.erl (hunk 2 above)
Comment (russelldb) [inline src/riak_core_stat_cache.erl]: Yeah, I only added it for the test really and cutandpasted it. `ok` will do fine.

File: src/riak_core_stat_cache.erl [hunk 5]
```diff
@@ -136,10 +167,19 @@ code_change(_OldVsn, State, _Extra) ->
...
 %% internal
-register_mod(App, Mod, Fun, Args, TTL, Apps) ->
+schedule_get_stats(After, App, MFA) ->
+    erlang:send_after(After, ?SERVER, {get_stats, {App, MFA}}).
+
+make_freshness_stat(App, TS) ->
+    {make_freshness_stat_name(App), TS}.
+
+make_freshness_stat_name(App) ->
+    list_to_atom(atom_to_list(App) ++ "_stat_ts").
```
Comment (rzezeski) [inline src/riak_core_stat_cache.erl:181]: Is there any reason we should worry about making stat names atoms?  Is there any way to users to cause stat names to be created?

Comment (russelldb) [inline src/riak_core_stat_cache.erl:181]: You have to register a module with riak_core:register to get to be a stat name. Or call riak_core_stat_cache:register. But the `App` must already be an atom. So we're taking an existing atom and adding "_ts" to it.

File: src/riak_core_stat_cache.erl (hunk 3 above)
Comment (russelldb) [inline src/riak_core_stat_cache.erl]: Ok,  I agree. It is messy and confusing. However, for the 1.3.1 release I don't want to change the `register_app` public interface to the `{{M, F, A}, RR}`, as that will mean changing all those mods that call that function, but from that boundary inwards I will change it to be consistent. Is that OK?

**Comment by russelldb** [comment]
Pooshed changes that address comments. Please to re-review.

File: src/riak_core_stat_cache.erl [hunk 6]
```diff
@@ -67,18 +80,22 @@ stop() ->
...
     process_flag(trap_exit, true),
     Tab = ets:new(?MODULE, [protected, set, named_table]),
-    TTL = app_helper:get_env(riak_core, stat_cache_ttl, ?TTL),
+    RefreshRateSecs = app_helper:get_env(riak_core, stat_cache_ttl, ?REFRESH_RATE),
+    RefreshRateMillis = ?REFRSH_MILLIS(RefreshRateSecs),
     %% re-register mods, if this is a restart after a crash
-    RegisteredMods = lists:foldl(fun({App, Mod}, Registerd) ->
-                                         register_mod(App, Mod, produce_stats, [], TTL, Registerd) end,
+    RegisteredMods = lists:foldl(fun({App, Mod}, Registered) ->
+                                         register_mod(App, ?DEFAULT_REG(Mod, RefreshRateMillis), Registered),
```
Comment (jrwest) [inline src/riak_core_stat_cache.erl]: think this needs to be:

```
Registered1 = register_mod(...),
schedule_get_stats(...)
Registered1
```

or something to that effect

Comment (jrwest) [inline src/riak_core_stat_cache.erl]: actually, seems there is a similar pattern below when handling `{register, App, Registraton}`. Any reason to not just roll `schedule_get_stats` into `register_mod`?

Comment (russelldb) [inline src/riak_core_stat_cache.erl]: Shit, yes it should, good catch.

I'll see about rolling schedule into `register_mod`

**Comment by jrwest** [comment]
code changes look good. tests pass (including `verify_riak_stats`). +1.

**Comment by russelldb** [comment]
Squashed and merged.

**Pull request closed without merging**